DM-44294: Initial queryDataIds implementation for RemoteButler

dhirving opened the pull request:
Add minimal support for DataCoordinate result type for the new query system, and shim RemoteButler.registry.queryIds to use it.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Add minimal implementation of query.data_ids

Add minimal support for DataCoordinate result type for the new query system

dhirving committed:
Factor out reusable query results code

Pull the portions of the DimensionRecordsQueryResults shim for the legacy query system that can be re-used for other result types.

dhirving committed:
Pull out an interface for DataCoordinateQueryResults

DataCoordinateQueryResults will need an implementation using the new query system, so factor out an abstract base class to serve as an interface.

dhirving committed:
Pull out common interface for legacy query results

To make the typing less convoluted for the new implementation, factor out an abstract base class that defines the common interface for the old query system results class.  This also removes some duplicated documentation.

dhirving committed:
Fix non-post-processed results being doubled

Fix an issue where results that did not require post-processing would return every row twice.  This was happening due to a missing return in a branch that tries to optimize the results by bypassing post-processing.

dhirving committed:
Handle query differences in testQueryDataIdsOrderBy

Update testQueryDataIdsOrderBy to deal with differences between the old query system and the new query system.  The old query system returns duplicate data ID rows in results, but the new one always de-duplicates.

dhirving committed:
Implement IN for new query system expression strings

dhirving committed:
Handle query system differences in testQueryResultSummaries

The new query system is able to give a more accurate count than the old query system, and is still missing some query diagnostics.

dhirving committed:
Add support for OVERLAPS operator in expression strings

These are already supported by the new query system and just needed to be forwarded from the expression string language.

dhirving committed:
Handle query system differences in testTimespanQueries

The new query system is intentionally missing a few timespan-related comparison operators that were supported by the old one.

dhirving committed:
Implement datetime overlaps timespan for new query

Implement "datetime OVERLAPS timespan" and "timespan OVERLAPS datetime" expression strings for the new query system.  These were supported by the old query system and are useful for things like checking whether an exposure time is in the embargo window.

Also added support for TimespanProxy.overlaps(astropy.time.Time), for use outside of expression strings.

dhirving committed:
Partially implement RemoteButler.registry.queryDataIds

Implement the basic functionality of queryDataIds for RemoteButler, leaving out the more complicated functions to derive new searches from the query results object.

dhirving committed:
Implement DataCoordinateQueryResults.subset for RemoteButler

dhirving committed:
Drop nonexistent queryDataIds feature in RemoteButler

It turns out registry.queryDataIds does not allow a 'None' value for the dimensions parameter, which is why there is no test coverage for this.  I think I got confused between registry.queryDataIds and the new query system's Query.data_ids, which does take a None.

TallJimbo commented on python/lsst/daf/butler/direct_query_driver/_driver.py:
```diff
@@ -1222,3 +1224,33 @@ def _convert_dimension_record_results(
+        Page object that holds a `DataCoordinate` container.
+    """
+    rows = [
+        DataCoordinate.standardize(cast(Mapping, row._mapping), dimensions=spec.dimensions)
```
`standardize` is full of checks and argument-wrangling we'd prefer to avoid here.  I tried to set things up such that
```
DataCoordinate.from_full_values(spec.dimensions, tuple(row[:len(spec.dimensions)])
```
will work, ultimately because that's the [iteration order for `ColumnSet`](https://github.com/lsst/daf_butler/blob/faa84923553689904c74f88840b8316e8f1de562/python/lsst/daf/butler/queries/tree/_column_set.py#L249-L251).  I'm not thrilled with that just being an unenforced behavioral requirement imposed on one piece of code by another piece of code several layers away, though, so I'm open to opinions on how to clean that up.

dhirving replied:
Yeah I don't think conceptually it's too terrible, since the `ResultSpec` we have available here owns the list of dimensions (`spec.dimensions`) and also owns the process of deciding what columns will be in the result (`spec.get_result_columns`).

Maybe I can add a method to ColumnSet (or ResultSpec-delegating-to-ColumnSet) that is equivalent to the `row[:len(spec.dimensions)]` bit of your code, and put it right next to `__iter__` so it's more local.  Or something similar to that -- basically explicitly keep track of which columns correspond to what somewhere.

dhirving replied:
How do you feel about this?: https://github.com/lsst/daf_butler/pull/1014/commits/d8a0a1bb9a772fa55a533722811bf04452ffbc78

(edited: the original commit didn't work because I didn't know the difference between DimensionGroup.names and DimensionGroup.data_coordinate_keys).

TallJimbo commented on python/lsst/daf/butler/remote_butler/server_models.py:
```diff
@@ -236,12 +236,35 @@ class QueryExecuteRequestModel(pydantic.BaseModel):
+    """
+
+    type: Literal["data_coordinate"] = "data_coordinate"
+    rows: list[SerializedDataCoordinate]
```
This is probably not the right time (I'd wait until we do `.expanded()`), but I think we're going to want this model to factor out the `DimensionGroup` that's common to all rows and store a `tuple` of values for each row, instead of using `SerializedDataCoordinate`.

I think we'll want something similar for dimension record results, too, but that's harder, because while `DataCoordinate` values are just `str` or `int`, `DimensionRecord` can have `Region` and `Timespan` and hence needs some kind of serialization logic involved.

dhirving replied:
Yeah these response models are basically just a "simplest thing that could possibly work" placeholder.  All the result data is eventually going to go through the pagination mechanism and I assume probably parquet or some other more efficient format.

TallJimbo reviewed (approved):
Looks great.  Only comments are probably-premature optimization things.

dhirving committed:
Fetch data ID columns by position instead of mapping

Added a helper class for finding columns in result rows.  Used it to pull out data ID columns by position instead of needing to needing to pass a mapping into DataCoordinate.standardize() -- Jim wanted to avoid the overhead.
